File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
k, n = map(int, input().split())
arr = [int(input()) for _ in range(k)]

def solve():
    result = 0

    arr.sort()
    l = 1
    r = arr[k - 1]

    while l <= r:
        mid = (l + r) // 2
        cut_size = cut_len_size(arr, mid)

        if cut_size < n:
            r = mid - 1
        else:
            l = mid + 1
            result = max(result, mid)

    return result

def cut_len_size(lens, mid):
    sum_of_len = 0
    for x in lens:
        sum_of_len += (x // mid)
    return sum_of_len

if __name__ == "__main__":
    print(solve())

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.StringTokenizer

fun main() {
    val (k, n, arr) = input()
    println(solve(k, n, arr))
}

private fun solve(k: Int, n: Int, arr: LongArray): Long {
    var l: Long = 1
    var r: Long = arr[k - 1]
    var answer: Long = 0

    while (l <= r) {
        val mid = ((l + r) / 2)
        if (cutLen(arr, mid) < n) {
            r = mid - 1
        } else {
            answer = mid
            l = mid + 1
        }
    }
    return answer
}

private fun cutLen(lens: LongArray, cuttingSize: Long): Long = lens.sumOf { (it / cuttingSize) }

private fun input(): Triple<Int, Int, LongArray> {
    val br = BufferedReader(InputStreamReader(System.`in`))

    val (k, n) = StringTokenizer(br.readLine()).run {
        nextToken().toInt() to nextToken().toInt()
    }

    val arr = LongArray(k) {
        br.readLine().toLong()
    }.apply { sort() }

    return Triple(k, n, arr)
}
